sql/rpl_rli.cc

```diff
@@ -438,6 +438,36 @@ bool Relay_log_info::mts_finalize_recovery() {
   DBUG_RETURN(ret);
 }
 
+bool Relay_log_info::mts_workers_queue_empty() {
+  ulong ret = 0;
+
+  for (Slave_worker **it = workers.begin(); it != workers.end(); ++it) {
+    mysql_mutex_lock(&(*it)->jobs_lock);
+    ret += (*it)->curr_jobs;
+    mysql_mutex_unlock(&(*it)->jobs_lock);
+  }
+
+  return ret == 0;
+}
+
+/* Checks if all in-flight stmts/trx can be safely rollbacked */
+bool Relay_log_info::cannot_safely_rollback() {
+  if (!is_parallel_exec())
+    return info_thd->get_transaction()->cannot_safely_rollback(
+        Transaction_ctx::SESSION);
+
+  bool ret = false;
+
+  for (Slave_worker **it = workers.begin(); it != workers.end(); ++it) {
+    mysql_mutex_lock(&(*it)->jobs_lock);
+    ret = ret || (*it)->info_thd->get_transaction()->cannot_safely_rollback(
+                     Transaction_ctx::SESSION);
```

Comment:
Why not just move bool ret to the loop scope, and break the loop if it's true, after the mutex unlock?